src/librustc/session/config.rs

```diff
@@ -2304,11 +2341,37 @@ pub fn build_session_options_and_crate_config(
             );
         };
 
+
+        // Externsl crates start out public,
+        // and become private if we later see
+        // an '--extern-private' key. They never
+        // go back to being public once we've seen
+        // '--extern-private', so we logical-AND
+        // their current and new 'public' value together
+
         externs
             .entry(name.to_owned())
             .or_default()
-            .insert(location);
-    }
+            .entry(location)
+            .and_modify(|e| *e &= public)
+            .or_insert(public);
+    }
+
+    // Now that we've determined the 'public' status of each extern,
+    // collect them into a set of ExternEntry
+    let externs: BTreeMap<String, BTreeSet<ExternEntry>> = externs.into_iter()
```

Comment:
With representation from https://github.com/rust-lang/rust/pull/59335/files#r271163300 the re-collection can be avoided.

Reply:
I'm pretty sure that each `location` is supposed to have its own `public` flag.

Reply:
Ah, I see the privacy is reset for duplicate paths, like `--extern name=my_path --extern-private name=my_path`.
I need to check what semantic the duplicates have in general, in the end `name` from source can refer only to a single paths (and a single privacy).

Reply:
I don't think that's the case. `private_dep` is stored in [CrateMetadata](https://github.com/rust-lang/rust/pull/59335/files#diff-974185a1d0122c6d970a1e4faaedbe5fR294), which is tied to a specific path.

Reply:
A single `CrateMetadata` corresponds to the whole `BTreeSet<ExternEntry>`.
Multiple externs with the same name
```
--extern name=PATH1 --extern name=PATH2
```
are allowed only in limited number of cases like
```
--extern name=rlib_path.rlib --extern name=dylib_path.dylib
```
(otherwise it's an error), and all dylib, rlib and rmeta components are combined together and written  into a single `CrateMetadata`, so privacies should be combined as well.

Actually, in addition to the `ExternEntry` change in https://github.com/rust-lang/rust/pull/59335#discussion_r271163300 I'd conservatively report an error during option parsing if `--extern`s with the same name have inconsistent privacies.